crypto/src/key/hdkd/derivation_path.rs

```diff
@@ -0,0 +1,248 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::key::hdkd::DerivationError;
+use crate::key::hdkd::DerivationError::*;
+use core::fmt;
+use std::fmt::Formatter;
+use std::slice::Iter;
+use std::str::FromStr;
+
+const PREFIX: &str = "m";
+const SEPARATOR: &str = "/";
+const HARDENED_BIT: u32 = 0x80000000;
+
+/// BIP32-like child numbers. Currently we only support hardened derivations
+#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd)]
+pub enum ChildNumber {
+    Hardened(u32),
+}
+
+impl ChildNumber {
+    /// Return a hardened child number
+    pub fn hardened(index: u32) -> Result<Self, DerivationError> {
+        if index & HARDENED_BIT == 0 {
+            Ok(ChildNumber::Hardened(index))
+        } else {
+            Err(InvalidChildNumber(index))
+        }
+    }
+
+    /// Return a normal child number. Currently unsupported
+    fn normal(index: u32) -> Result<Self, DerivationError> {
+        if index & HARDENED_BIT == 0 {
+            // For the time being we don't support non-hardened derivations
+            Err(UnsupportedDerivationType)
+        } else {
+            Err(InvalidChildNumber(index))
+        }
+    }
+
+    /// Return a BIP32-like child number index that has the hardened bit set if needed
+    pub fn to_encoded_index(&self) -> u32 {
+        match self {
+            ChildNumber::Hardened(i) => i | HARDENED_BIT,
+        }
+    }
+
+    /// Returns true if this child number is hardened
+    pub fn is_hardened(&self) -> bool {
+        match self {
+            ChildNumber::Hardened(_) => true,
+        }
+    }
+}
+
+impl FromStr for ChildNumber {
+    type Err = DerivationError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // Check if this child number is hardened
```

Comment:
Please treat these strictly as ascii bytes. Like I mentioned before, we should not support any unicode or other weird encoding string that works just by coincidence. There's no reason to support it. This is a recipe for problems. This whole wallet system is based on ascii. Let's try to follow that for simplicity.

The correct way to do this is to start by converting this to bytes, check that all the bytes are ascii, and only then start processing.

I do recognize that you added tests for specific cases, but I don't think we can test all possibilities because encoding is a mess in general.

I don't know if I'm missing something here that makes all this OK. But I have to be cautious.

Reply:
As far as I can tell, the parsing algorithm already only accepts ASCII strings. The trick is to specify what is accepted, rather than what is rejected, which already happens. The only possible exception might be third-party code like `u32::parse` which is used here. I expect it only accepts ASCII too but may be worth double checking.

Reply:
I just checked with the stdlib implementation of `u32::parse` and it only accepts ASCII chars. However, it seems it still allows the plus sign even in front of an unsigned number. That means we would accept strings like `"m/+42'"` (please check). Do we want this?

Reply:
Nice catch! Nope, + isn't allowed.

Reply:
One thing that also occurred to me... what about exponents? would something like 1e6 parse and return 1000000? That would be very ridiculous! 😄

Quite honestly, this is why I'm overly cautious with parsing strings that are meant to be read as bytes... there's always that corner case that you didn't think of... always some exotic case that'll cause trouble.

In C++, I would use `isdigit()` to check every character after the suffix as bytes. I'm sure there's an equivalent thing in Rust.

Reply:
> One thing that also occurred to me... what about exponents? would something like 1e6 parse and return 1000000? That would be very ridiculous! smile

Heh. No, from what I've seen, only `+` was there. Also, it may allow leading zeroes.

Reply:
The idea was that the `parse` to `u32` would fail on non ascii characters. Having said that, I don't have any issues with checking if the `str` is ascii. Better safe than sorry.